src/mast/mod.rs

```diff
@@ -0,0 +1,1108 @@
+    pub fn nest(&mut self) {
+        self.current_scope += 1;
+    }
+
+    /// Exits a scoped block.
+    pub fn pop(&mut self) {
+        self.current_scope = self.current_scope.checked_sub(1).expect("scope bug");
+
+        // remove variables as we exit the scope
+        let mut to_delete = vec![];
+        for (name, (scope, _)) in self.vars.iter() {
+            if *scope > self.current_scope {
+                to_delete.push(name.clone());
+            }
+        }
+        for d in to_delete {
+            self.vars.remove(&d);
+        }
+    }
+
+    /// Returns true if a scope is a prefix of our scope.
+    pub fn is_in_scope(&self, prefix_scope: usize) -> bool {
+        self.current_scope >= prefix_scope
+    }
+
+    /// Stores type information about a local variable.
+    pub fn store_type(&mut self, ident: &str, type_info: &MTypeInfo) -> Result<()> {
+        match self
+            .vars
+            .insert(ident.to_string(), (self.current_scope, type_info.clone()))
+        {
+            Some(_) => Err(Error::new(
+                "type-checker",
+                ErrorKind::DuplicateDefinition(ident.to_string()),
+                type_info.span,
+            )),
+            None => Ok(()),
+        }
+    }
+
+    /// Retrieves type information on a variable, given a name.
+    /// If the variable is not in scope, return None.
+    pub fn get_type_info(&self, ident: &str) -> Option<&MTypeInfo> {
+        if let Some((scope, type_info)) = self.vars.get(ident) {
+            if self.is_in_scope(*scope) {
+                Some(type_info)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// A context to store the last node id for the monomorphized AST.
+#[derive(Debug, Default)]
+pub struct MastCtx {
+    pub last_node_id: usize,
+}
+
+impl MastCtx {
+    pub fn new() -> Self {
+        Self { last_node_id: 0 }
+    }
+
+    pub fn next_node_id(&mut self) -> usize {
+        self.last_node_id += 1;
+        self.last_node_id
+    }
+}
+
+#[derive(Debug)]
+/// Mast relies on the TAST for the information about the "unresolved" types to monomorphize.
+/// Things such as loading the function AST and struct AST from fully qualified names.
+/// After monomorphization process, the following data will be updated:
+/// - Resolved types. This can be used to determine the size of a type.
+/// - Instantiated functions. The circuit writer will load the instantiated function AST by node id.
+/// - Monomorphized AST is generated for the circuit writer to walk through and compute.
+pub struct Mast<B>
+where
+    B: Backend,
+{
+    tast: TypeChecker<B>,
+
+    /// Mapping from node id to resolved type
+    node_types: HashMap<usize, TyKind>,
+
+    /// Mapping from node id to instantiated functions
+    node_functions: HashMap<usize, FnInfo<B>>,
+
+    ctx: MastCtx,
+
+    /// Monomorphized AST of the main function
+    pub main_fn_ast: Option<FunctionDef>,
```

Comment:
> I'm guessing this is because we assign a new node id to each nodes

After monomorphized, the resulted AST will be different:
- generic types will be resolved as concrete types
- each function instantiation will create a new AST (with generic parameter resolved) for the `FnCall` expression node to point to.
- thus, the node id will need to be regenerated.

Reply:
In terms of the use of Option, I think it is same reason for the `node_types` and `node_functions` fields. So if they don't get init via the Mast::new(), then they need to be passed around through the process of monomorphization. 

Alternatively, maybe we can make the `main_fn_ast` private and expose via functions so as to encapsulate the downsides of the use of Option.

Reply:
can't you do

```rust
fn monomorphize(tast: TAST) -> MAST {
  let main_fn_ast = /* ... */;
  let ctx = /* ... * /;
  /* monomorphize... */ 
  MAST {
    main_fn_ast,
    // etc.
  }
}
```

Reply:
Attempting to fix this at [7e02fef](https://github.com/zksecurity/noname/pull/147/commits/7e02fef4cba1a371893755f249dfcdefa85857bf)

Because the monomophi* functions are moved out of the Mast struct, the change of the indent of the code leads to a huge amount of git diff...

Reply:
you may want to check the diffs in vscode, which highlight the diffs "smarter way". I think the github highlight for these diffs can be overwhelming TT

Reply:
it's alright I need to rereview this whole file anyway :o